### PR TITLE
new features

### PR DESCRIPTION
**Pull request by rainbreak**
No compiler errors or warnings with 0.4.23, 0.4.26, 0.5.17, 0.6.12, 0.7.5. Uses the latest hevm features.

### CODE REVIEW

File: src/test.sol [hunk 1]
```diff
@@ -121,7 +160,125 @@ contract DSTest {
...
+    function assertGt(int a, int b) internal {
+        if (a <= b) {
+            emit log_bytes32("Error: a > b not satisfied");
+            emit log_named_int("         a", a);
+            emit log_named_int("         b", b);
+            fail();
+        }
+    }
+    function assertGe(uint a, uint b, bytes32 err) internal {
+        if (a <= b) {
```
Comment (MrChico) [inline src/test.sol:192]:
```suggestion
        if (a < b) {
```

File: src/test.sol [hunk 2]
```diff
@@ -14,32 +16,38 @@
...
+    event log_named_string       (string key, string val);
+
+    bool public IS_TEST = true;
     bool public failed;
 
-    constructor() internal {
-        IS_TEST = true;
-    }
+    address constant HEVM_ADDRESS =
+        address(bytes20(uint160(uint256(keccak256('hevm cheat code')))));
```
Comment (MrChico) [inline src/test.sol:42]: I think the `bytes20` is redundant here
```suggestion
        address(uint160(uint256(keccak256('hevm cheat code'))));
```

**Review by MrChico — APPROVED** [review]

File: src/test.sol (hunk 2 above)
Comment (rainbreak) [reply src/test.sol:42]: Until solc tightens up type conversion even further :upside_down_face:

Comment (MrChico) [reply src/test.sol:42]: it seems like its a noop here so its fine, but in general I'm wary of bytes20 because it rightpads

File: src/test.sol [hunk 3]
```diff
@@ -14,32 +16,38 @@
...
+
+    modifier mayRevert() { _; }
+    modifier testopts(string memory) { _; }
```
Comment (xwvvvvwx) [inline src/test.sol:45]: Should we leave these out until we actually support them in `hevm`?

File: src/test.sol [hunk 4]
```diff
@@ -14,32 +16,38 @@
 pragma solidity >=0.4.23;
 
 contract DSTest {
-    event eventListener          (address target, bool exact);
+    event log                    (string);
     event logs                   (bytes);
+
+    event log_address            (bytes32);
```
Comment (xwvvvvwx) [inline src/test.sol:22]:
```suggestion
    event log_address            (address);
```

File: src/test.sol [hunk 5]
```diff
@@ -49,81 +57,356 @@ contract DSTest {
...
+    function assertLtDecimal(uint a, uint b, uint decimals) internal {
+        if (a >= b) {
+            emit log("Error: a < b not satisfied [decimal uint]");
+            emit log_named_decimal_uint("  Value a", a, decimals);
+            emit log_named_decimal_uint("  Value b", b, decimals);
+            fail();
+        }
+    }
+    function assertLtDecimal(uint a, uint b, uint decimals, string memory err) internal {
+        if (a <= b) {
```
Comment (xwvvvvwx) [inline src/test.sol:330]:
```suggestion
        if (a >= b) {
```

File: src/test.sol (hunk 3 above)
Comment (rainbreak) [reply src/test.sol:45]: imo no because upgrading ds-test is a pain

File: src/test.sol (hunk 5 above)
Comment (rainbreak) [reply src/test.sol:330]: :grimacing: :laughing:

File: src/test.sol [hunk 6]
```diff
@@ -133,12 +416,19 @@ contract DSTest {
         } else {
             ok = false;
         }
-
-        if (!ok) {
-            emit log_bytes32("Error: Wrong `bytes' value");
-            emit log_named_bytes32("  Expected", "[cannot show `bytes' value]");
-            emit log_named_bytes32("  Actual", "[cannot show `bytes' value]");
+    }
+    function assertEq0(bytes memory a, bytes memory b) internal {
```
Comment (xwvvvvwx) [inline src/test.sol:420]: why can't this be `assertEq0`?

Comment (rainbreak) [reply src/test.sol:420]: It clashes with `assertEq(string,string)`

Comment (rainbreak) [reply src/test.sol:420]: oh, or it used to hmm

Comment (xwvvvvwx) [reply src/test.sol:420]: oh they have the same abi encoding?

Comment (rainbreak) [reply src/test.sol:420]: you'll get an error like 

```
Error: No unique declaration found after argument-dependent lookup.
   --> aux/demo.sol:212:9:
    |
212 |         assertEq(hex"abcdef01", hex"abcdef02");
    |         ^^^^^^^^
```

**Review by xwvvvvwx — APPROVED** [review]

**Comment by kaycoinsUSA** [comment]
Done

**Pull request merged**